### PR TITLE
People Management: Display Followers

### LINKED ISSUE #5174: People: Display Site's Readers
Building on the people management work done thus far, allow the user to display a site's followers:

3w8mcis-ta-900x900

ref: #5030

### PR DESCRIPTION
**Pull request by jleandroperez**
#### Description:
In this PR we're implementing the capability to view a Website's Followers.

Closes #5174

#### Details:
- Although **Users** and **Followers* come from different endpoints, delta between the two data structures is minimum.
- In order to reuse code, as much as possible, **Users** and **Followers** will be cached in the same Core Data store.
- We're also reusing the `View Users` machinery, since the UI is exactly the same.

--

#### Scenario: View Users
1. Log into your WordPress.com account
2. Open `My Sites` > `Site` > `People` (you've need to be an admin)

Please, verify that:
- The site's Users must be listed. 
- User's details (for any entry that's not your own) should allow you to update the Role, and delete the user.

#### Scenario: View Followers
1. Log into your WordPress.com account
2. Open `My Sites` > `Site` > `People` (you've need to be an admin)
3. Tap over the Navigation Bar, and pick `Followers`

Please, verify that:
- The site's Followers must be listed. 
- Follower's details shouldn't allow to update the Role, nor to delete the entity.

#### Scenario: Update User Role
1. Log into your WordPress.com account
2. Open `My Sites` > `Site` > `People` (you've need to be an admin)
3. Tap over the Role row, and make any change.

Please, verify that the change gets reflected in WordPress.com

#### Scenario: Nuke User
1. Log into your WordPress.com account
2. Open `My Sites` > `Site` > `People` (you've need to be an admin)
3. Tap over the `Remove User` row.

Please, verify that the change gets reflected in WordPress.com

--

Needs review: @diegoreymendez 

@diegoreymendez i'll defer to your good taste sir!. I've reused the `People Management` code, as much as possible, and would love to have your opinion on this one.

As an alternative to this PR, i've considered adding a new Entity (`Follower`), and another set of ViewControllers (`FollowersViewController` / `FollowerViewController`), but i found myself duplicating `PeopleViewController` / `PersonViewController`.

Insights / feedback, as always, will be more than welcome.

Thanks in advance!

### CODE REVIEW

**Comment by diegoreymendez** [comment]
As discussed by PM with @jleandroperez:

The approach seemed fine to me at first, the PR shows some very nice work.

As we discussed it though, we ended up agreeing that mixing site-members and site-followers could complicate things a bit.  This is because these two entities share most of their fields, but they don't share IDs.

So while it's possible to create a primary key that involves the ID + some other field, we felt it was undesirable from the point of view of maintainability.

So long story short: @jleandroperez will try separating these into two entities sharing a common base entity.  The details will be defined in his upcoming commits.

**Comment by jleandroperez** [comment]
@diegoreymendez first improvements iteration involves:

- Maintaining a single Core Data entity, for simplicity: **ManagedPerson**
- A **Person** protocol, implemented by two structs, **User** and **Follower**
- Generics here and there
- A cleaner **PeopleService** API

Please, lemme know your thoughts. Any kind of feedback will be more than welcome, no filters please 😄 !!!

**Comment by jleandroperez** [comment]
P.s.: Test Cases updated, to account for all of the services that got updated.

File: WordPress/Classes/Networking/PeopleRemote.swift [hunk 1]
```diff
@@ -185,10 +208,14 @@ private extension PeopleRemote {
     /// - Parameters:
     ///     - response: Raw backend dictionary
     ///     - siteID: the ID of the site associated
+    ///     - type: The kind of Person we should parse.
     ///
     /// - Returns: A single *Person* instance.
     ///
-    private func personFromResponse(user: [String: AnyObject], siteID: Int) throws -> Person {
+    func personFromResponse<T : Person>(user   : [String: AnyObject],
```
Comment (diegoreymendez) [inline WordPress/Classes/Networking/PeopleRemote.swift]: The alignment of the ":" is off, when comparing this line to the following ones.

We should seriously consider not having to do manual vertical-alignment. 😈

File: WordPress/Classes/ViewRelated/People/PersonViewController.swift [hunk 2]
```diff
@@ -135,7 +135,7 @@ final class PersonViewController : UITableViewController {
         roleViewController.blog = blog
         roleViewController.selectedRole = person.role
         roleViewController.onChange = { newRole in
-            self.updateRole(newRole)
+            self.updateUserRole(newRole)
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/People/PersonViewController.swift]: Storing a strong reference to a parent to a child is not very recommendable, as it's almost sure the parent has a strong reference to the child as well (thus creating a retain cycle).

Since changing `self` to be `weak` here adds no risk to the code at hand, I'd consider doing so.

File: WordPress/Classes/ViewRelated/People/PersonViewController.swift [hunk 3]
```diff
@@ -177,25 +177,35 @@ private extension PersonViewController {
...
+            self.deleteUser()
         }
 
         alert.presentFromRootViewController()
     }
 
-    func removePerson() {
+    func deleteUser() {
+        guard let user = user else {
+            DDLogSwift.logError("Error: Only Users can be deleted")
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/People/PersonViewController.swift]: We can keep the log for release builds, but for debug builds this should be an assertion failure since it clearly means something's wrong with the code logic.

File: WordPress/Classes/ViewRelated/People/PersonViewController.swift [hunk 4]
```diff
@@ -177,25 +177,35 @@ private extension PersonViewController {
...
         let service = PeopleService(blog: blog)
-        service?.deletePerson(person)
+        service?.deleteUser(user)
         navigationController?.popViewControllerAnimated(true)
     }
 
-    func updateRole(newRole: Person.Role) {
+    func updateUserRole(newRole: Role) {
+        guard let user = user else {
+            DDLogSwift.logError("Error: Only Users have Roles!")
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/People/PersonViewController.swift]: Same as above... if we're logging an error it means this should be an assertion failure in debug builds.

**Comment by diegoreymendez** [comment]
@jleandroperez - All working great.  Just a few comments on the code.  Let me know.

File: WordPress/Classes/ViewRelated/People/PersonViewController.swift (hunk 3 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/People/PersonViewController.swift]: Good call, adding. Thanks!

File: WordPress/Classes/ViewRelated/People/PersonViewController.swift (hunk 4 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/People/PersonViewController.swift]: Adding an assert here, as well. Thank you!

File: WordPress/Classes/Networking/PeopleRemote.swift (hunk 1 above)
Comment (jleandroperez) [inline WordPress/Classes/Networking/PeopleRemote.swift]: Alignment gone, sorry about that!... that's my OCD hitting buttons 😄

File: WordPress/Classes/ViewRelated/People/PersonViewController.swift (hunk 2 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/People/PersonViewController.swift]: Updating, thanks!

**Comment by jleandroperez** [comment]
All comments addressed @diegoreymendez!

Ready for another look, thank you!

File: WordPress/Classes/ViewRelated/People/PersonViewController.swift [hunk 5]
```diff
@@ -176,26 +176,38 @@ private extension PersonViewController {
...
+        alert.addDestructiveActionWithTitle(removeTitle) { [weak self] action in
+            self?.deleteUser()
         }
 
         alert.presentFromRootViewController()
     }
 
-    func removePerson() {
+    func deleteUser() {
+        assert(user != nil)
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/People/PersonViewController.swift]: This is not really wrong at all, so feel free to ignore this comment.  It's very subjective.

I was thinking that maybe it'd make sense to just add `assertionFailure()` inside the guard below.  Up to you.

**Comment by diegoreymendez** [comment]
@jleandroperez - Added a very subjective comment, so feel free to ignore it.

Looking great! :shipit:

**Comment by jleandroperez** [comment]
Thanks a lot @diegoreymendez !!
